### PR TITLE
cephfs: Add documentation about snapshot create/restore and clone

### PR DESCRIPTION
**Pull request by humblec**
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

updates: #1188

### CODE REVIEW

**Comment by Madhu-1** [comment]
@humblec lets track the PRs and issues with milestone, not labels.

File: docs/cephfs-snap-clone.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+### Snapshot PVCs
+
+```bash
+## create snapshotclass
+$kubectl create -f snapshotclass.yaml
+
+## create snapshot
+
+$ kubectl create -f snapshot.yaml
```
Comment (Madhu-1) [inline docs/cephfs-snap-clone.md:32]: please add a link to cephfs readme for how to create PVC and also mention that PVC should be already created to create a snapshot

File: docs/cephfs-snap-clone.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+volumesnapshot.snapshot.storage.k8s.io/cephfs-pvc-snapshot created
+
+## Get details about the snapshot
+
+$ kubectl get volumesnapshot
+NAME                  READYTOUSE   SOURCEPVC    SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS                SNAPSHOTCONTENT                                    CREATIONTIME   AGE
+cephfs-pvc-snapshot   true         cephfs-pvc                           1Gi           csi-cephfsplugin-snapclass   snapcontent-34476204-a14a-4d59-bfbc-2bbba695652c   3s             6s
+
+$ kubectl describe volumesnapshot cephfs-pvc-snapshot
```
Comment (Madhu-1) [inline docs/cephfs-snap-clone.md:41]: why do we need to describe it?  this adds more content to file makes it hard to read

File: docs/cephfs-snap-clone.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+$
+
+## Get details about the volumesnapshotcontent
+
+$ kubectl get volumesnapshotcontent
+NAME                                               READYTOUSE   RESTORESIZE   DELETIONPOLICY   DRIVER                          VOLUMESNAPSHOTCLASS          VOLUMESNAPSHOT        AGE
+snapcontent-34476204-a14a-4d59-bfbc-2bbba695652c   true         1073741824    Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-snapclass   cephfs-pvc-snapshot   64s
+$
+
+$ kubectl describe volumesnapshotcontent snapcontent-34476204-a14a-4d59-bfbc-2bbba695652c
```
Comment (Madhu-1) [inline docs/cephfs-snap-clone.md:103]: same here as well

File: docs/cephfs-snap-clone.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+## Restore snapshot to a new PVC :
+
+$ kubectl create -f pvc-restore.yaml
+
+$ kubectl get pvc
+NAME                 STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
+cephfs-pvc           Bound    pvc-1ea51547-a88b-4ab0-8b4a-812caeaf025d   1Gi        RWO            rook-cephfs    20h
+cephfs-pvc-restore   Bound    pvc-95308c75-6c93-4928-a551-6b5137192209   1Gi        RWO            rook-cephfs    11m
+$ kubectl describe pv pvc-95308c75-6c93-4928-a551-6b5137192209
```
Comment (Madhu-1) [inline docs/cephfs-snap-clone.md:191]: describe outputs are not needed here as we are not explaining anything in it.

File: docs/cephfs-snap-clone.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+  - [create snapshot](#create-snapshot)
+  - [Get details about the snapshot](#get-details-about-the-snapshot)
+  - [Get details about the volumesnapshotcontent](#get-details-about-the-volumesnapshotcontent)
+  - [Restore snapshot to a new PVC :](#restore-snapshot-to-a-new-pvc-)
+    - [CLONE PVCs](#clone-pvcs)
+  - [Get details about the cloned PVC](#get-details-about-the-cloned-pvc)
+
+## Prerequisite
+
+- For snapshot functionality to be supported for your kubernetes cluster, the
```
[the remark this answers is not in the document]
Comment (humblec) [reply docs/cephfs-snap-clone.md:16]: done.

File: docs/cephfs-snap-clone.md [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+  kubernetes version running in your cluster should be >= v1.17. We also need the
```
[the remark this answers is not in the document]
Comment (humblec) [reply docs/cephfs-snap-clone.md:17]: done.thanks..

File: docs/cephfs-snap-clone.md [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+  `snapshot controller` deployed in your kubernetes cluster along with `csi-snapshotter`
+  sidecar container. These sidecar controllers watch for snasphot requests and push the
```
[the remark this answers is not in the document]
Comment (humblec) [reply docs/cephfs-snap-clone.md:19]: sure. done.

File: docs/cephfs-snap-clone.md [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+  requests to Ceph CSI driver for snapshot creation,deletion operations. There should
+  be a `volumesnapshotclass` object present in the cluster for snapshot request to be satisified.
+
+
+### Snapshot PVCs
+
+```bash
```
[the remark this answers is not in the document]
Comment (humblec) [reply docs/cephfs-snap-clone.md:26]: done.

File: docs/cephfs-snap-clone.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (humblec) [reply docs/cephfs-snap-clone.md:32]: done

Comment (humblec) [reply docs/cephfs-snap-clone.md:32]: done.

**Comment by humblec** [comment]
@madhu-1 ptal .

**Comment by humblec** [comment]
ah.. line length issues.. solving it atm.

File: docs/cephfs-snap-clone.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+cephfs-pvc-snapshot   true         csi-cephfs-pvc                          1Gi           csi-cephfsplugin-snapclass   snapcontent-34476204-a14a-4d59-bfbc-2bbba695652c   3s             6s
+```
+
+- Get details about the volumesnapshotcontent
+
+```console
+$ kubectl get volumesnapshotcontent
+NAME                                               READYTOUSE   RESTORESIZE   DELETIONPOLICY   DRIVER                          VOLUMESNAPSHOTCLASS          VOLUMESNAPSHOT        AGE
+snapcontent-34476204-a14a-4d59-bfbc-2bbba695652c   true         1073741824    Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-snapclass   cephfs-pvc-snapshot   64s
+$
```
[the remark this answers is not in the document]
Comment (humblec) [reply docs/cephfs-snap-clone.md]: @nixpanic  removed extra `$`. Regarding one instance of `$kubectl` it was kept that way because there is no output to that command . Without out md linter will complain :+1:

File: docs/cephfs-snap-clone.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  for more information on these sidecar controllers. There should
+  be a `volumesnapshotclass` object present in the cluster
+  for snapshot request to be satisified.
+
+## Snapshot
+
+### Create snapshotclass
+
+```console
+$kubectl create -f snapshotclass.yaml
```
[the remark this answers is not in the document]
Comment (humblec) [reply docs/cephfs-snap-clone.md]: added the path to the file @nixpanic . thanks.

**Comment by humblec** [comment]
@nixpanic @Madhu-1 ptal .

File: docs/cephfs-snap-clone.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  for more information on these sidecar controllers. There should
+  be a `volumesnapshotclass` object present in the cluster
+  for snapshot request to be satisified.
+
+## Snapshot
+
+### Create snapshotclass
+
+```console
+$kubectl create -f ../examples/cephfs/snapshotclass.yaml
```
Comment (Madhu-1) [inline docs/cephfs-snap-clone.md:27]: add space between $ and kubelet

File: examples/cephfs/pvc-clone.yaml [hunk 12]
```diff
@@ -9,7 +9,7 @@ spec:
     name: csi-cephfs-pvc
     kind: PersistentVolumeClaim
   accessModes:
-    - ReadWriteOnce
+    - ReadWriteMany
```
Comment (Madhu-1) [inline examples/cephfs/pvc-clone.yaml:12]: why this change is here?

Comment (humblec) [reply examples/cephfs/pvc-clone.yaml:12]: ah. removing it.

File: docs/cephfs-snap-clone.md (hunk 11 above)
Comment (humblec) [reply docs/cephfs-snap-clone.md:27]: @Madhu-1  as in previous comment https://github.com/ceph/ceph-csi/pull/1343#discussion_r468533502

File: docs/cephfs-snap-clone.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (Madhu-1) [reply docs/cephfs-snap-clone.md]: in that case, we don't need to use $ see https://github.com/ceph/ceph-csi/tree/master/examples

[the remark this answers is not in the document]
Comment (humblec) [reply docs/cephfs-snap-clone.md:56]: sure. Done.

**Comment by humblec** [comment]
@Madhu-1 @nixpanic ptal . comments are addressed. :+1:

File: docs/cephfs-snap-clone.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+  for more information on these sidecar controllers. There should
+  be a `volumesnapshotclass` object present in the cluster
+  for snapshot request to be satisified.
+
+## Snapshot
+
+### Create snapshotclass
+
+```console
+kubectl create -f snapshotclass.yaml
```
Comment (nixpanic) [inline docs/cephfs-snap-clone.md:27]: Not sure where my comment went... I asked where this file comes from, and if you can explain that in the document.

File: docs/cephfs-snap-clone.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (nixpanic) [reply docs/cephfs-snap-clone.md:56]: Sorry, I don't get it. `$` is still there, and there are still inconsistencies with `kubectl`, `$ kubectl` (this document) and `[$] kubectl` (from other docs).

File: docs/cephfs-snap-clone.md (hunk 13 above)
Comment (humblec) [reply docs/cephfs-snap-clone.md:27]: Ah.. I missed it while removing the other commit .. sorry.. bringing it back.

Comment (humblec) [reply docs/cephfs-snap-clone.md]: @nixpanic ptal .. restored that version :)

File: docs/cephfs-snap-clone.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (nixpanic) [reply docs/cephfs-snap-clone.md:56]: This was marked as resolved, but I do not think it is?

**Comment by humblec** [comment]
@Madhu-1 can you bring back the approval ?

File: docs/cephfs-snap-clone.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+cephfs-pvc-snapshot   true         csi-cephfs-pvc                          1Gi           csi-cephfsplugin-snapclass   snapcontent-34476204-a14a-4d59-bfbc-2bbba695652c   3s             6s
+```
+
+- Get details about the volumesnapshotcontent
+
+```console
+$ kubectl get volumesnapshotcontent
+NAME                                               READYTOUSE   RESTORESIZE   DELETIONPOLICY   DRIVER                          VOLUMESNAPSHOTCLASS          VOLUMESNAPSHOT        AGE
+snapcontent-34476204-a14a-4d59-bfbc-2bbba695652c   true         1073741824    Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-snapclass   cephfs-pvc-snapshot   64s
+$
```
Comment (Madhu-1) [inline docs/cephfs-snap-clone.md:56]: this is not fixed yet @humblec I see extra $

File: docs/cephfs-snap-clone.md [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+cephfs-pvc-restore   Bound    pvc-95308c75-6c93-4928-a551-6b5137192209   1Gi        RWO            rook-cephfs    11m
+```
+
+## Clone PVC
+
+```console
+$ kubectl create -f ../examples/cephfs/pvc-clone.yaml
+$ kubectl get pvc
+NAME                 STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
+csi-cephfs-pvc          Bound    pvc-1ea51547-a88b-4ab0-8b4a-812caeaf025d   1Gi        RWO            rook-cephfs    20h
```
Comment (Madhu-1) [inline docs/cephfs-snap-clone.md:76]: alignment is not proper here

File: docs/cephfs-snap-clone.md [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+```
+
+## Restore snapshot to a new PVC
+
+```console
+
+$ kubectl create -f ../examples/cephfs/pvc-restore.yaml
+$ kubectl get pvc
+NAME                 STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
+csi-cephfs-pvc          Bound    pvc-1ea51547-a88b-4ab0-8b4a-812caeaf025d   1Gi        RWO            rook-cephfs    20h
```
Comment (Madhu-1) [inline docs/cephfs-snap-clone.md:66]: same here

File: docs/cephfs-snap-clone.md [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+NAME                  READYTOUSE   SOURCEPVC    SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS                SNAPSHOTCONTENT                                    CREATIONTIME   AGE
+cephfs-pvc-snapshot   true         csi-cephfs-pvc                          1Gi           csi-cephfsplugin-snapclass   snapcontent-34476204-a14a-4d59-bfbc-2bbba695652c   3s             6s
+```
+
+- Get details about the volumesnapshotcontent
+
+```console
+$ kubectl get volumesnapshotcontent
+NAME                                               READYTOUSE   RESTORESIZE   DELETIONPOLICY   DRIVER                          VOLUMESNAPSHOTCLASS          VOLUMESNAPSHOT        AGE
+snapcontent-34476204-a14a-4d59-bfbc-2bbba695652c   true         1073741824    Delete           rook-ceph.cephfs.csi.ceph.com   csi-cephfsplugin-snapclass   cephfs-pvc-snapshot   64s
```
Comment (Madhu-1) [inline docs/cephfs-snap-clone.md:55]: why rook driver is used here?

File: docs/cephfs-snap-clone.md (hunk 15 above)
Comment (Madhu-1) [inline docs/cephfs-snap-clone.md:76]: the storageclass name needs to be changed. currently, it's pointing to rook not storage class of cephcsi  examples

**Comment by humblec** [comment]
@Madhu-1 comments are addressed. ptal :+1:

File: docs/cephfs-snap-clone.md [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+```console
+$ kubectl create -f ../examples/cephfs/snapshot.yaml
+volumesnapshot.snapshot.storage.k8s.io/cephfs-pvc-snapshot created
+```
+
+- Get details about the snapshot
+
+```console
+$ kubectl get volumesnapshot
+NAME                  READYTOUSE   SOURCEPVC    SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS                SNAPSHOTCONTENT                                    CREATIONTIME   AGE
```
Comment (Madhu-1) [inline docs/cephfs-snap-clone.md:46]: this alignment looks to be broken https://github.com/ceph/ceph-csi/blob/5e305ab65cb51f0cc345ddaa0ec449dc0d816b1f/docs/cephfs-snap-clone.md#create-snapshot

Comment (humblec) [reply docs/cephfs-snap-clone.md:46]: @Madhu-1 can you refresh https://github.com/ceph/ceph-csi/pull/1343/commits/5e305ab65cb51f0cc345ddaa0ec449dc0d816b1f?short_path=9b1252c#diff-9b1252c8f53ed336e48e29788d07810f ?

Comment (humblec) [reply docs/cephfs-snap-clone.md:46]: I see.. one sec

Comment (humblec) [reply docs/cephfs-snap-clone.md:46]: @Madhu-1 should be good now

File: docs/cephfs-snap-clone.md [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+NAME                                               READYTOUSE   RESTORESIZE   DELETIONPOLICY   DRIVER                VOLUMESNAPSHOTCLASS          VOLUMESNAPSHOT        AGE
+snapcontent-34476204-a14a-4d59-bfbc-2bbba695652c   true         1073741824    Delete           cephfs.csi.ceph.com   csi-cephfsplugin-snapclass   cephfs-pvc-snapshot   64s
+```
+
+## Restore snapshot to a new PVC
+
+```console
+
+$ kubectl create -f ../examples/cephfs/pvc-restore.yaml
+$ kubectl get pvc
```
Comment (Madhu-1) [inline docs/cephfs-snap-clone.md:63]: sorry @humblec I forgot on the last thing all cephfs PVC access mode is RWX in examples

Comment (humblec) [reply docs/cephfs-snap-clone.md:63]: @Madhu-1 changing everything to RWX

Comment (humblec) [reply docs/cephfs-snap-clone.md:63]: done.

**Comment by humblec** [comment]
@Mergify rebase

**Comment by Madhu-1** [comment]
@MergifyIO rebase

**Comment by humblec** [comment]
@mergify rebase

**Comment by nixpanic** [comment]
> /test commitlint

This actually failed, see https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/commitlint/43/ But, the status was not reported back. I'll restart the test, it will fail again, but at least there will be a result posted and Mergify can continue.

#1372 most likely fixes the CI job, but it needs to get merged first, and then this can be rebased

**Comment by Madhu-1** [comment]
@mergifyIO rebase
